File: mobile/test/java/ru/yandex/yandexbus/inhouse/mvp/BasePresenterTest.java
package ru.yandex.yandexbus.inhouse.mvp;

import org.junit.Before;
import org.junit.Test;
import rx.Subscription;

import java.util.NoSuchElementException;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

public class BasePresenterTest {
    private interface TestView {}

    private BasePresenter<TestView> presenter;
    private TestView view;

    @Before
    public void beforeEachTest() {
        view = new TestView() {};
        presenter = new BasePresenterImplDelegate<>();
    }

    @Test
    public void bindView_shouldAttachViewToThePresenter() {
        presenter.bind(view);
        assertEquals(presenter.view(), view);
    }

    @Test(expected = IllegalStateException.class)
    public void bindView_shouldThrowIfPreviousViewIsNotUnbounded() {
        presenter.bind(view);
        presenter.bind(createView());
    }

    @Test(expected = NoSuchElementException.class)
    public void view_shouldThrowByDefault() {
        presenter.view();
    }

    @Test
    public void view_shouldReturnViewAfterBind() {
        presenter.bind(view);
        assertEquals(presenter.view(), view);
    }

    @Test
    public void unsubscribeOnUnbindView_shouldWorkAccordingItsContract() {
        presenter.bind(view);

        Subscription subscription1 = mock(Subscription.class);
        Subscription subscription2 = mock(Subscription.class);
        Subscription subscription3 = mock(Subscription.class);

        presenter.unsubscribeOnUnbind(subscription1, subscription2, subscription3);
        verify(subscription1, never()).unsubscribe();
        verify(subscription2, never()).unsubscribe();
        verify(subscription3, never()).unsubscribe();

        presenter.unbind(view);
        verify(subscription1).unsubscribe();
        verify(subscription2).unsubscribe();
        verify(subscription3).unsubscribe();
    }

    @Test(expected = NoSuchElementException.class)
    public void unbindView_shouldNullTheViewReference() {
        presenter.bind(view);
        presenter.unbind(view);
        presenter.view();
    }

    @Test(expected = IllegalStateException.class)
    public void unbindView_shouldThrowIfPreviousViewIsNotSameAsExpected() {
        presenter.bind(view);
        presenter.unbind(createView());
    }

    private TestView createView() {
        return new TestView() {};
    }
}

File: mobile/test/java/ru/yandex/yandexmaps/common/mvp/BasePresenterTest.kt
package ru.yandex.yandexmaps.common.mvp

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown
import org.junit.Before
import org.junit.Test

class BasePresenterTest {
    private interface TestView

    private lateinit var presenter: BasePresenterWithViewValidation<TestView>
    private lateinit var view: TestView

    @Before
    fun beforeEachTest() {
        view = object : TestView {}
        presenter = object : BasePresenterWithViewValidation<TestView>() {}
    }

    @Test
    fun bindView_shouldAttachViewToThePresenter() {
        presenter.bind(view)
        assertThat(presenter.view()).isSameAs(view)
    }

    @Test
    fun bindView_shouldThrowIfPreviousViewIsNotUnbounded() {
        presenter.bind(view)
        try {
            presenter.bind(createView())
            failBecauseExceptionWasNotThrown(IllegalStateException::class.java)
        } catch (expected: IllegalStateException) {
            assertThat(expected).hasMessage("Previous view is not unbound! previousView = $view")
        }
    }

    @Test
    fun view_shouldThrowByDefault() {
        try {
            presenter.view()
            failBecauseExceptionWasNotThrown(NullPointerException::class.java)
        } catch (expected: IllegalStateException) {
            assertThat(expected).hasMessage("View is null. You should probably bind before accessing view!")
        }
    }

    @Test
    fun view_shouldReturnViewAfterBind() {
        presenter.bind(view)
        assertThat(presenter.view()).isSameAs(view)
    }

    @Test
    fun accessViewAfterUnbind_shouldThrow() {
        presenter.bind(view)
        presenter.unbind(view)
        try {
            presenter.view()
            failBecauseExceptionWasNotThrown(NullPointerException::class.java)
        } catch (expected: IllegalStateException) {
            assertThat(expected).hasMessage("View is null. You should probably bind before accessing view!")
        }
    }

    @Test
    fun unbindView_shouldThrowIfPreviousViewIsNotSameAsExpected() {
        presenter.bind(view)
        val unexpectedView = createView()
        try {
            presenter.unbind(unexpectedView)
            failBecauseExceptionWasNotThrown(IllegalStateException::class.java)
        } catch (expected: IllegalStateException) {
            assertThat(expected).hasMessage("Unexpected view! previousView = $view, view to unbind = $unexpectedView")
        }
    }

    private fun createView(): TestView {
        return object : TestView {}
    }
}
